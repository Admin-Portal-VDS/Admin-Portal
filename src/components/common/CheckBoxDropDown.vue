<template>
  <div class="checkbox-dropdown" @click.stop>
    <button class="dropdown-button" @click="toggleDropdown">
      {{ checkBoxDropDownTitle }}
      <span class="arrow" :class="{ open: isOpen }"></span>
    </button>

    <div v-if="isOpen" class="dropdown-menu">
      <div class="dropdown-item">
        <label>
          <input
            type="checkbox"
            :checked="isSelectAllChecked"
            @change="toggleSelectAll"
            data-id="select-all"
          />
          Select All
        </label>
      </div>
      <div
        v-for="(option, index) in dropDownList"
        :key="index"
        class="dropdown-item"
      >
        <label>
          <input
            type="checkbox"
            :value="option.key"
            v-model="selectedOptions"
            :data-id="`option-${index}`"
          />
          {{ option.label }}
        </label>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    checkBoxDropDownTitle: {
      type: String,
      required: true,
    },
    dropDownList: {
      type: Array,
      required: true,
    },
  },
  data() {
    return {
      isOpen: false,
      selectedOptions: [],
    };
  },
  computed: {
    isSelectAllChecked() {
      const allKeys = this.dropDownList.map((option) => option.key);
      return (
        allKeys.length > 0 &&
        allKeys.every((key) => this.selectedOptions.includes(key))
      );
    },
  },
  methods: {
    toggleDropdown(event) {
      event.stopPropagation();
      this.isOpen = !this.isOpen;
    },
    closeDropdown() {
      this.isOpen = false;
    },
    toggleSelectAll() {
      if (this.isSelectAllChecked) {
        this.selectedOptions = [];
      } else {
        this.selectedOptions = this.dropDownList.map((option) => option.key);
      }
      this.$emit('update-selected', this.selectedOptions);
    },
  },
  mounted() {
    document.addEventListener('click', this.closeDropdown);
  },
  beforeUnmount() {
    document.removeEventListener('click', this.closeDropdown);
  },
};
</script>

<style scoped>
.checkbox-dropdown {
  position: relative;
  display: inline-block;
}

.arrow {
  display: inline-block;
  width: 5px;
  height: 5px;
  margin-left: 10px;
  border: solid black;
  border-width: 0 1px 1px 0;
  transform: rotate(45deg);
  transition: transform 0.3s ease;
}

.arrow.open {
  transform: rotate(-135deg);
}

.dropdown-button {
  border: 1.3px solid;
  border-radius: 6px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  font-size: 0.9rem;
  font-weight: 590;
  font-style: normal;
  text-align: center;
  padding: 10px 15px;
  cursor: pointer;
  min-height: 30px;
  min-width: 50px;
  background-color: transparent;
  color: #070000;
  white-space: nowrap;
}

.dropdown-menu {
  position: absolute;
  top: 40px;
  left: 0;
  background-color: white;
  border: 1px solid #ccc;
  border-radius: 6px;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
  z-index: 1000;
  padding: 10px;
  min-width: 120px;
}

.dropdown-item {
  margin: 5px 0;
  padding: 8px 10px;
  font-size: 14px;
  font-weight: 350;
}

label {
  display: flex;
  align-items: center;
  gap: 5px;
}

input[type='checkbox'] {
  cursor: pointer;
  accent-color: black;
  border-radius: 35px;
  padding: 5px;
  border: 4px solid black;
  background-color: #222;
  transition: all 0.3s ease;
  display: block;
  flex: 0 0 22px;
  height: 22px;
  margin-right: 10px;
  margin-top: -1px;
  position: relative;
  width: 22px;
}
</style>
