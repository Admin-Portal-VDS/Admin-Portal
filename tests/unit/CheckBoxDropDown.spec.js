import { mount } from '@vue/test-utils';
import CheckBoxDropDown from '@/components/common/CheckBoxDropDown';

describe('CheckboxDropdown.vue', () => {
  let wrapper;
  const title = 'Select Options';
  const options = [
    { key: 'user', label: 'User' },
    { key: 'extension', label: 'Extension' },
    { key: 'phone number', label: 'Phone Number' },
  ];

  beforeEach(() => {
    wrapper = mount(CheckBoxDropDown, {
      props: {
        checkBoxDropDownTitle: title,
        dropDownList: options,
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('renders the dropdown with the correct title', () => {
    expect(wrapper.find('[data-id="dropdown-button"]').text()).toContain(title);
  });

  it('toggles the dropdown menu visibility when button is clicked', async () => {
    expect(wrapper.find('[data-id="dropdown-menu"]').exists()).toBe(false);
    await wrapper.find('[data-id="dropdown-button"]').trigger('click');
    expect(wrapper.find('[data-id="dropdown-menu"]').exists()).toBe(true);
    await wrapper.find('[data-id="dropdown-button"]').trigger('click');
    expect(wrapper.find('[data-id="dropdown-menu"]').exists()).toBe(false);
  });

  it('renders the correct number of checkboxes for options', async () => {
    await wrapper.find('[data-id="dropdown-button"]').trigger('click');
    const checkboxes = wrapper.findAll('input[type="checkbox"]');
    expect(checkboxes.length).toBe(options.length + 1); //+1 for select all
  });

  it('updates selectedOptions when individual checkboxes are checked', async () => {
    await wrapper.find('[data-id="dropdown-button"]').trigger('click');
    const firstCheckbox = wrapper.find('[data-id="option-0"]');
    await firstCheckbox.setChecked();
    expect(wrapper.vm.selectedOptions).toContain('user');
  });

  it('checks "Select All" checkbox when all options are selected', async () => {
    await wrapper.find('[data-id="dropdown-button"]').trigger('click');
    const checkboxes = wrapper.findAll('input[type="checkbox"]');
    await checkboxes[1].setChecked();
    await checkboxes[2].setChecked();
    await checkboxes[3].setChecked();
    expect(wrapper.vm.isSelectAllChecked).toBe(true);
    expect(checkboxes[0].element.checked).toBe(true);
  });

  it('toggles all checkboxes when "Select All" is clicked', async () => {
    await wrapper.find('[data-id="dropdown-button"]').trigger('click');
    const selectAllCheckbox = wrapper.find('[data-id="select-all"]');
    await selectAllCheckbox.setChecked();
    expect(wrapper.vm.selectedOptions).toEqual(options.map((opt) => opt.key));
    await selectAllCheckbox.setChecked(false);
    expect(wrapper.vm.selectedOptions).toEqual([]);
  });

  it('emits "update-selected" event with selected options', async () => {
    await wrapper.find('[data-id="dropdown-button"]').trigger('click');
    const selectAllCheckbox = wrapper.find('[data-id="select-all"]');
    await selectAllCheckbox.setChecked();
    expect(wrapper.emitted('update-selected')).toBeTruthy();
    expect(wrapper.emitted('update-selected')[0][0]).toEqual(
      options.map((opt) => opt.key),
    );
  });

  it('closes the dropdown when clicking outside', async () => {
    await wrapper.find('[data-id="dropdown-button"]').trigger('click');
    expect(wrapper.vm.isOpen).toBe(true);
    document.dispatchEvent(new Event('click'));
    expect(wrapper.vm.isOpen).toBe(false);
  });

  it('accepts props and passes them correctly', () => {
    const wrapper = mount(CheckBoxDropDown, {
      props: { checkBoxDropDownTitle: title, dropDownList: options },
    });

    expect(wrapper.props('checkBoxDropDownTitle')).toBe(title);
    expect(wrapper.props('dropDownList')).toStrictEqual(options);
  });

  it('matches snapshot', () => {
    const wrapper = mount(CheckBoxDropDown, {
      propsData: {
        checkBoxDropDownTitle: 'Snapshot Test',
        dropDownList: options,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
