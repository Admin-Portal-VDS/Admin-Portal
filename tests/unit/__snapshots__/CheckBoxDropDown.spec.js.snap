// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxDropdown.vue matches snapshot 1`] = `
<div class="checkbox-dropdown"><button class="dropdown-button">Snapshot Test <span class="arrow"></span></button>
  <!--v-if-->
</div>
`;
